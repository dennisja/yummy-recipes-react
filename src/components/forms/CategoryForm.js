import React, {Component} from 'react';
import { PropTypes } from 'prop-types';

import CategoryRequest from '../../helpers/Categories';


class AddCategoryForm extends Component {
    static initialState = {
        cat_name: ''
    }

    static defaultProps = {
        role: 'add'
    }
    
    state = AddCategoryForm.initialState

    componentDidMount () {
        if(this.props.role === 'edit'){
            CategoryRequest.getCategory(this.props.categoryId)
            .then((response)=>{
                //update state with the current name of the category to fill up the category name
                this.setState({
                    cat_name: response.data.recipe_cat.name,
                })
            })
            .catch((error)=>{
                //if cateogory doent exist, redirect and flag an error
                if(error.response){
                    console.log(error.response)
                }else if(error.request){
                    console.log(error.request)
                }
            });
        }
    }
    
    
    handleInputChange = (event)=>{
        const target = event.target;
        this.setState({
            [target.name]:target.value
        })
    }


    handleCategorySubmit = (event)=>{
        event.preventDefault();
        const {role} = this.props;

        if(role === 'add'){
            //add recipe 
            CategoryRequest.addCategory(this.state)
            .then(response=>{
                console.log(response)
            })
            .catch(error=>{
                console.log(error)
            })
        }

        if(role==='edit'){
            //edit category
            CategoryRequest.editCategory(this.props.categoryId, this.state)
            .then(response=>{
                window.Materialize.toast("Successfully Edited category", 4000);
                this.props.history.push('/categories')
                console.log(response)
            })
            .catch(error=>{
                if(error.response){
                    console.log(error);
                }else if(error.request){
                    console.log(error);
                }
            })
        }
    }

    render() {
        const {cat_name} = this.state;
        const {role} = this.props;

        return (
            <form onSubmit={this.handleCategorySubmit}>
                <div className="input-field">
                    <label htmlFor="cat_name">Category Name:</label>
                    <input type="text" name="cat_name" id="cat_name" value={cat_name} onChange={this.handleInputChange} required/>
                </div>
                <div className="center-align">
                    <input
                        type='submit'
                        id='submit_category'
                        className="btn btn-small orange"
                        name="submit_category"
                        value={`${role === 'add' ? 'Add' : 'Edit'} Category`}/>
                </div>
            </form>
        );
    }
}

AddCategoryForm.propTypes = {
    role: PropTypes.string,
}

export default AddCategoryForm;