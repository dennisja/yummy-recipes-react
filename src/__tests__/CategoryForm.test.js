import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import CategoryForm from '../components/forms/CategoryForm';

describe('Test CategoryForm', () => {
    const formWrapper = shallow(<CategoryForm />);
    const preventDefault = jest.fn();


    it('renders properly', () => {
        expect(shallowToJson(formWrapper)).toMatchSnapshot();
    });

    it('renders a form element', () => {
        expect(formWrapper.find('form')).toHaveLength(1);
    });

    it('has a name field and submit button', () => {
        expect(formWrapper.find('#cat_name')).toHaveLength(1);
        expect(formWrapper.find('#submit_category')).toHaveLength(1);
    });

    it('submits data', () => {
        formWrapper.setState({ cat_name: 'Hello' });
        formWrapper.find('form').simulate('submit', { preventDefault });
        expect(preventDefault).toBeCalled();
        formWrapper.find('#submit_category').simulate('click');
    });

    it('renders edit button when handling edit', () => {
        formWrapper.setProps({ role: 'edit' });
        formWrapper.update();
        // check whethert it has the input field
        expect(formWrapper.find('#cat_name')).toHaveLength(1);
        expect(formWrapper.find('#submit_category')).toHaveLength(1);
        // check value of submit button
        expect(formWrapper.find('#submit_category').props().value).toBe('Edit Category');
    });

    it('handles editing category', () => {
        formWrapper.setProps({ role: 'edit' });
        expect(formWrapper.instance().props.role).toEqual('edit');
        expect(formWrapper.find('form')).toHaveLength(1);
        formWrapper.find('form').simulate('submit', { preventDefault });
        expect(preventDefault).toBeCalled();
    });
});
