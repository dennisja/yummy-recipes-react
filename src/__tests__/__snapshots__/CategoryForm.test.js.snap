// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test CategoryForm renders properly 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="input-field"
  >
    <label
      htmlFor="cat_name"
    >
      Category Name:
    </label>
    <input
      id="cat_name"
      name="cat_name"
      onChange={[Function]}
      required={true}
      type="text"
      value=""
    />
  </div>
  <div
    className="center-align"
  >
    <input
      className="btn btn-small orange"
      id="submit_category"
      name="submit_category"
      type="submit"
      value="Add Category"
    />
  </div>
</form>
`;
